cwltool/singularity.py

```diff
@@ -79,39 +86,55 @@ def get_image(dockerRequirement,  # type: Dict[Text, Text]
 
         if "dockerImageId" not in dockerRequirement and "dockerPull" in dockerRequirement:
             match = re.search(pattern=r'([a-z]*://)', string=dockerRequirement["dockerPull"])
-            candidate = _normalize_image_id(dockerRequirement['dockerPull'])
-            candidates.append(candidate)
-            dockerRequirement['dockerImageId'] = candidate
+            candidate_image = _normalize_image_id(dockerRequirement['dockerPull'])
+            candidates.append(candidate_image)
+            candidate_sif = _normalize_sif_id(dockerRequirement['dockerPull'])
+            candidates.append(candidate_sif)
```

Comment:
This is currently only being added to `candidates[]` for searching but this could also be used in the future if singularity version is specified. For example, the .sif file is 3x specific, so something like if `if version == '3.x': candidate = _normalize_sif_id; else: candidate = _normalize_image_id`

Reply:
I think we established that 2.x uses `.img` and 3.x uses `.sif`?

Reply:
We did, but there is no code which checks for singularity version at runtime. I currently have the code checking for both .img and .sif because of this. In the future we can choose if we'd like to search only for .sif or .img depending on singularity version.